controllers/rollingupgrade_controller.go

```diff
@@ -489,6 +513,21 @@ func (r *RollingUpgradeReconciler) runRestack(ctx *context.Context, ruObj *upgra
 	return processedInstances, nil
 }
 
+func (r *RollingUpgradeReconciler) resolveLaunchDefinition(asg *autoscaling.Group) (*launchDefinition, error) {
```

Comment:
Does it make more sense to have a single return with both fields?
```
return &launchDefinition{
			kind:                    launchKindLaunchConfiguration,
			launchConfigurationName: asg.LaunchConfigurationName,
                         launchTemplate: asg.LaunchTemplate,
```
If it's Nil, it will stay Nil anyway :) and then you don't have duplicate code.
If you need the context of which one it is downstream, you can have helper predicates `isLaunchConfig()` and `isLaunchTemplate()`

Reply:
ASGs can also use [MixedInstancesPolicy](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_MixedInstancesPolicy.html), so, need to make sure code can be easily extended later.

Reply:
Can you elaborate on how `MixedInstancesPolicy` might change things? you can still have a single return with all required fields and check if specific ones are nil downstream if it's required.
Also, we might want to store the values so it's easier later e.g. `launchConfigurationName: aws.StringValue(asg.LaunchConfigurationName)`

Reply:
@eytan-avisror I've added support for `MixedInstancesPolicy` too, turned out we use it in some clusters already.